neighsyncd/neighsyncd.cpp

```diff
@@ -4,15 +4,34 @@
 #include "netdispatcher.h"
 #include "netlink.h"
 #include "neighsyncd/neighsync.h"
+#include "warm_restart.h"
+#include <time.h>
 
 using namespace std;
 using namespace swss;
 
 int main(int argc, char **argv)
 {
     Logger::linkToDbNative("neighsyncd");
-    DBConnector db(APPL_DB, DBConnector::DEFAULT_UNIXSOCKET, 0);
-    NeighSync sync(&db);
+
+    DBConnector appDb(APPL_DB, DBConnector::DEFAULT_UNIXSOCKET, 0);
+    RedisPipeline pipelineAppDB(&appDb);
+    DBConnector confDb(CONFIG_DB, DBConnector::DEFAULT_UNIXSOCKET, 0);
+    RedisPipeline pipelineConfDB(&confDb);
+
+    WarmStart::checkWarmStart("neighsyncd");
+
+    NeighSync sync(&pipelineAppDB, &pipelineConfDB);
+
+    if (sync.m_warmStartInProgress)
+    {
+        WarmStart::setWarmStartState("neighsyncd", WarmStart::INIT);
+        // To-be-discussed: Do we need flush the producestate table pipeline to
+        // make sure the relevant Table has the up-to-date data before this daemon restarted?
+        // (e,g, pending messages were not writen to the Table)
+        // On the other hand, in case docker restart, we need make sure the flush
+        // won't break anything for orchagent
```

Comment:
For planned warm restart, maybe pipleline flush could be done before shutdown.  What is the pipeline batch number for neighsyncd?   If not 1,  it looks neighsyncd should be process the pre-shutdown state check signal too.

The one that is more tricky is how to deal with those entries which have been created by producestate but not read by consumerstate yet.   Maybe before reading table from APPDB,  no pending data should exist in producestate buffer.

Reply:
fpmsyncd has the same issues to deal with during bgp docker warm restart:  pipeline flush and buffered set/del operation in producerstate set. 

It looks we need to make orchagent_restart_check program in https://github.com/Azure/sonic-swss/pull/562 a common tool beyond orchagent.

Reply:
just to conclude here, pipeline flush is not necessary for pre-shutdown, as long as the consumer data and database plain table data are in sync, we should be fine as we are going to get the new table from application/kernel again. 

After daemon is restarted, we need make sure the producerstate table is empty or cleared, see:
https://github.com/Azure/sonic-swss-common/pull/227